ark/phenotyping/som_utils.py

```diff
@@ -67,6 +68,136 @@ def compute_cluster_avg(fovs, channels, base_dir, cluster_col,
+
+
+def compute_cell_cluster_counts(fovs, channels, base_dir, consensus_dir,
+                                cell_table_path, cluster_col='cluster'):
+    """Create a matrix with each fov-cell label pair and their pixel SOM/meta cluster counts
+
+    Args:
+        fovs (list):
+            The list of fovs to subset on
+        channels (list):
+            The list of channels to subset on
+        base_dir (str):
+            The path to the data directories
+        consensus_dir (str):
+            Name of directory to save the consensus clustered results
+        cell_table_path (str):
+            Path to the cell table, needs to be created with Segment_Image_Data.ipynb
+        cluster_col (str):
+            The name of the pixel cluster column to count per cell,
+            should be 'cluster' or 'hCluster_cap'
+
+    Returns:
+        pd.DataFrame:
+            cell x cluster list of counts of each pixel SOM/meta cluster per each cell
+    """
+
+    # define the counts per segmentation label DataFrame
+    cluster_counts = pd.DataFrame()
+
+    # read the cell table data
+    cell_table = pd.read_csv(cell_table_path)
+
+    # subset on fov, label, and cell size
+    cell_table = cell_table[['fov', 'label', 'cell_size']]
+
+    # for verification purposes, make sure the fovs are sorted in numerical order
+    fovs_sorted = sorted(fovs, key=lambda x: int(re.findall(r'\d+', x)[0]))
+
+    for fov in fovs_sorted:
+        # read the full dataset with pixel and meta clusters
+        fov_pixel_data = feather.read_dataframe(
+            os.path.join(base_dir, consensus_dir, fov + '.feather')
+        )
+
+        # discard fov as we'll get it from the cell table
+        fov_pixel_data = fov_pixel_data.drop(columns='fov')
+
+        # group the data by segmentation label and cluster value
+        group_by_cluster_col = fov_pixel_data.groupby(
+            ['segmentation_label', cluster_col]
+        ).size().reset_index(name='count')
+
+        # cast the cluster column to int for aesthetics
+        group_by_cluster_col[cluster_col] = group_by_cluster_col[cluster_col].astype(int)
+
+        # create a pivot table with counts of each cluster per segmentation label
+        num_cluster_per_seg_label = group_by_cluster_col.pivot(
+            index='segmentation_label', columns=cluster_col, values='count'
+        ).fillna(0).astype(int)
+
+        # change column names to indicate which cluster type (pixel or meta) is used
+        new_columns = ['%s_' % cluster_col + str(c) for c in num_cluster_per_seg_label.columns]
+        num_cluster_per_seg_label.columns = new_columns
+
+        # copy over the segmentation label data
+        num_cluster_per_seg_label['segmentation_label'] = num_cluster_per_seg_label.index.values
+
+        # reset the index
+        num_cluster_per_seg_label = num_cluster_per_seg_label.reset_index(drop=True)
+
+        # subset cell table on specific fov, convert labels to int to match data types
```

Comment:
@ngreenwald to clarify, instead of going through all the trouble of subsetting `cell_table` just merge it directly in with `num_cluster_per_seg_label`?

Reply:
Yeah, if you can merge on 'fov' and 'label' column to get a unique match, you should be able to merge the num_cluster_per_seg_label directly into 'cell_table'. This would simplify what's going on I think